Add support for the Strimzi Metrics Reporter to Kafka Connect and MirrorMaker2

OwenCorrigan76 opened the pull request:
### Type of change

- Enhancement / new feature


### Description
This patch adds support for the Strimzi Metrics Reporter to Kafka Connect and MirrorMaker2 as described by the following proposal:

https://github.com/strimzi/proposals/blob/main/064-prometheus-metrics-reporter.md

Related to #10753

We won’t initially support the CruiseControl component. To make it work, CC should be changed to expose metrics through its HTTP endpoint.

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

OwenCorrigan76 committed:
Add tests to KafkaConnectConfigurationBuilderTest

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
Address Fede's comments and add tests

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
Rebase and prep for PR

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -269,20 +274,50 @@ private void printConfigProviders(AbstractConfiguration userConfig) {
+    public KafkaConnectConfigurationBuilder withUserConfiguration(AbstractConfiguration configurations,
+                                                                  boolean injectKafkaJmxReporter,
+                                                                  boolean injectStrimziMetricsReporter) {
+        if (configurations instanceof KafkaConnectConfiguration) {
```
```suggestion
        // creating a defensive copy to avoid mutating the input config
        if (configurations instanceof KafkaConnectConfiguration) {
```

OwenCorrigan76 replied:
Added

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -269,20 +274,50 @@ private void printConfigProviders(AbstractConfiguration userConfig) {
+        } else if (configurations instanceof KafkaMirrorMaker2Configuration) {
+            configurations = new KafkaMirrorMaker2Configuration((KafkaMirrorMaker2Configuration) configurations);
+        } else {
+            configurations = new KafkaConnectConfiguration(Reconciliation.DUMMY_RECONCILIATION, new ArrayList<>());
```
This is not a test. You can't use dummy reconciliation, but you need to pass the actual reconciliation as first parameter in the constructor of this class and then use it here.

OwenCorrigan76 replied:
Is this correct now?

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -269,20 +274,50 @@ private void printConfigProviders(AbstractConfiguration userConfig) {
+            ModelUtils.createOrAddListConfig(userConfig, "consumer.metric.reporters", values);
+        }
+        if (injectStrimziMetricsReporter) {
+            String values = "kafka.metrics.KafkaPrometheusMetricsReporter";
```
Did you forget the `io.strimzi.` prefix here?

OwenCorrigan76 replied:
Added

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -311,6 +346,23 @@ public KafkaConnectConfigurationBuilder withPluginPath() {
+            printSectionHeader("Strimzi Metrics Reporter configuration");
+            writer.println("prometheus.metrics.reporter.listener.enable=true");
+            writer.println("prometheus.metrics.reporter.listener=http://:" + StrimziMetricsReporterModel.METRICS_PORT);
+            writer.println("prometheus.metrics.reporter.allowlist=" + reporterModel.getAllowList());
```
You also have to repeat these configurations with `admin.`, `producer.` and `consumer.` prefixes.

OwenCorrigan76 replied:
Added

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMaker2Configuration.java:
```diff
@@ -34,6 +34,16 @@ public class KafkaMirrorMaker2Configuration extends AbstractConfiguration {
+    /**
+     * Constructor used to add user configuration.
+     *
+     * @param configuration User provided MirrorMaker 2 configuration
```
```suggestion
     * Copy constructor which creates new instance of the Mirror Maker 2 configuration from existing configuration.
     * It is useful when you need to modify an instance of the configuration without permanently changing the original.
     *
     * @param configuration     Existing configuration
```

OwenCorrigan76 replied:
Refactored

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfiguration.java:
```diff
@@ -33,6 +33,17 @@ public class KafkaConnectConfiguration extends AbstractConfiguration {
     }
 
+    /**
+     * Constructor used to add user configuration.
+     *
+     * @param configuration User provided Kafka Connect configuration
```
```suggestion
     * Copy constructor which creates new instance of the Kafka Connect configuration from existing configuration.
     * It is useful when you need to modify an instance of the configuration without permanently changing the original.
```

OwenCorrigan76 replied:
Refactored

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectClusterTest.java:
```diff
@@ -2164,18 +2166,34 @@ public void testMetricsParsingFromConfigMap() {
-        assertThat(kc.metrics().getConfigMapName(), is(nullValue()));
-        assertThat(kc.metrics().getConfigMapKey(), is(nullValue()));
+        assertThat(kc.metrics(), is(notNullValue()));
+        assertThat(((StrimziMetricsReporterModel) kc.metrics()).getAllowList(), is("kafka_log.*,kafka_network.*"));
```
We miss network policy test here:

```java
NetworkPolicy np = kc.generateNetworkPolicy(true, null, null);
List<NetworkPolicyIngressRule> rules = np.getSpec().getIngress().stream()
            .filter(ing -> ing.getPorts().get(0).getPort().equals(new IntOrString(StrimziMetricsReporterModel.METRICS_PORT)))
            .collect(Collectors.toList());

assertThat(rules.size(), is(1));
```

OwenCorrigan76 replied:
Is it in the correct place now?

fvaleri replied:
Nope, just append that code last, after:

```
assertThat(((StrimziMetricsReporterModel) kc.metrics()).getAllowList(), is("kafka_log.*,kafka_network.*"));
```

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -398,7 +402,7 @@ public void testWithRackId() {
     public void testWithConfigProviders() {
         String configuration = new KafkaConnectConfigurationBuilder(BOOTSTRAP_SERVERS)
-                .withUserConfigurations(null)
+                .withUserConfiguration(null, true, true)
```
This should be all false to preserve the existing behavior.

OwenCorrigan76 replied:
Done

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -424,7 +438,7 @@ public void testWithUserProvidedAndDefaultConfigurations() {
 
         String configuration = new KafkaConnectConfigurationBuilder(BOOTSTRAP_SERVERS)
-                .withUserConfigurations(configurations)
+                .withUserConfiguration(configurations, true, true)
```
This should be all false to preserve the existing behavior.

OwenCorrigan76 replied:
Done

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -458,7 +476,7 @@ public void testWithUserProvidedConfigMaps() {
 
         String configuration = new KafkaConnectConfigurationBuilder(BOOTSTRAP_SERVERS)
-                .withUserConfigurations(configurations)
+                .withUserConfiguration(configurations,  false, false)
```
```suggestion
                .withUserConfiguration(configurations, false, false)
```

OwenCorrigan76 replied:
Done

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -514,4 +532,217 @@ public void withPluginPath() {
     }
+
+    @ParallelTest
+    public void testNullUserConfiguration()  {
```
We don't need this test as user config can't be null.

OwenCorrigan76 replied:
Removed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -514,4 +532,217 @@ public void withPluginPath() {
+    }
+
+    @ParallelTest
+    public void testUserConfigurationWithConfigProviders() {
```
We don't need this test as it is covered by the parametric test with the permutations.

OwenCorrigan76 replied:
Removed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -514,4 +532,217 @@ public void withPluginPath() {
+    }
+
+    @ParallelTest
+    public void testNullUserConfigurationWithStrimziMetricsReporter() {
```
We don't need this test as it is covered by the parametric test with the permutations.

OwenCorrigan76 replied:
Removed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -514,4 +532,217 @@ public void withPluginPath() {
+    }
+
+    @ParallelTest
+    public void testNullUserConfigurationWithJmxMetricsReporter() {
```
We don't need this test as it is covered by the parametric test with the permutations.

OwenCorrigan76 replied:
Removed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -514,4 +532,217 @@ public void withPluginPath() {
+    static Stream<Arguments> userConfigurationWithMetricsReporters() {
+        Map<String, Object> configMap = new HashMap<>();
+        //configMap.put("auto.create.topics.enable", "false");
+        configMap.put("metric.reporters", "my.domain.CustomMetricReporter");
```
Where are all the other configurations with prefixes?

fvaleri replied:
This is still not correct. We are only interested in `metric.reporters` behavior here.

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilderTest.java:
```diff
@@ -514,4 +532,217 @@ public void withPluginPath() {
+        Map<String, Object> configMap = new HashMap<>();
+        //configMap.put("auto.create.topics.enable", "false");
+        configMap.put("metric.reporters", "my.domain.CustomMetricReporter");
+        configMap.put("kafka.metrics.reporters", "my.domain.CustomYammerMetricReporter");
```
This configuration does not exists on Kafka clients.

OwenCorrigan76 replied:
You mean the ones to add to `configmap` ?

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    }
+
+    @ParallelTest
+    public void testCreateOrAddConfigListWithDuplicate() {
```
```suggestion
    public void testCreateOrAddListConfigWithDuplicates() {
```

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    }
+
+    @ParallelTest
+    public void testCreateOrAddConfigListOrdering() {
```
```suggestion
    public void testCreateOrAddListConfigOrdering() {
```

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    }
+
+    @ParallelTest
+    public void testCreateOrAddConfigListWithNullConfig() {
```
```suggestion
    public void testCreateOrAddListConfigWithNullConfig() {
```

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    }
+
+    @ParallelTest
+    public void testCreateOrAddConfigListWithNullKey() {
```
```suggestion
    public void testCreateOrAddListConfigWithNullKey() {
```

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    }
+
+    @ParallelTest
+    public void testCreateOrAddConfigListWithNullValue() {
```
```suggestion
    public void testCreateOrAddListConfigWithNullValue() {
```

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    public void testCreateOrAddConfigListWithNullValue() {
+        KafkaConfiguration config = new KafkaConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, java.util.Set.of());
+        config.setConfigOption("test-key", "test-value-1,test-value-2");
+        java.util.stream.Stream.of(null, "", " ")
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    public void testCreateOrAddConfigListWithNullKey() {
+        KafkaConfiguration config = new KafkaConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, java.util.Set.of());
+        config.setConfigOption("test-key", "test-value-1,test-value-2");
+        java.util.stream.Stream.of(null, "", " ")
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    @ParallelTest
+    public void testCreateOrAddListConfigDoesNotExists() {
+        Map<String, Object> userConfiguration = new java.util.HashMap<>();
+        KafkaConnectConfiguration config1 = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+        ModelUtils.createOrAddListConfig(config1, "test-key", "test-value");
+        assertThat(config1.getConfigOption("test-key"), is("test-value"));
+
+        KafkaConnectConfiguration config2 = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    @ParallelTest
+    public void testCreateOrAddListConfigExists() {
+        Map<String, Object> userConfiguration = new java.util.HashMap<>();
+        KafkaConnectConfiguration config1 = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+        ModelUtils.createOrAddListConfig(config1, "test-key", "test-value-2");
+        assertThat(config1.getConfigOption("test-key"), is("test-value-1,test-value-2"));
+
+        KafkaConnectConfiguration config2 = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    @ParallelTest
+    public void testCreateOrAddListConfigExistsAndContainsValue() {
+        Map<String, Object> userConfiguration = new java.util.HashMap<>();
+        KafkaConnectConfiguration config1 = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+        ModelUtils.createOrAddListConfig(config1, "test-key", "test-value-1");
+        assertThat(config1.getConfigOption("test-key"), is("test-value-1"));
+
+        KafkaConnectConfiguration config2 = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    @ParallelTest
+    public void testCreateOrAddConfigListWithDuplicate() {
+        Map<String, Object> userConfiguration = new java.util.HashMap<>();
+        KafkaConnectConfiguration config = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    @ParallelTest
+    public void testCreateOrAddConfigListWithDuplicate() {
+        Map<String, Object> userConfiguration = new java.util.HashMap<>();
+        KafkaConnectConfiguration config = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    @ParallelTest
+    public void testCreateOrAddConfigListOrdering() {
+        Map<String, Object> userConfiguration = new java.util.HashMap<>();
+        KafkaConnectConfiguration config = new KafkaConnectConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, userConfiguration.entrySet());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+
+    @ParallelTest
+    public void testCreateOrAddConfigListWithNullConfig() {
+        KafkaConfiguration config = new KafkaConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, java.util.Set.of());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+
+    @ParallelTest
+    public void testCreateOrAddConfigListWithNullKey() {
+        KafkaConfiguration config = new KafkaConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, java.util.Set.of());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+    public void testCreateOrAddConfigListWithNullKey() {
+        KafkaConfiguration config = new KafkaConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, java.util.Set.of());
+        config.setConfigOption("test-key", "test-value-1,test-value-2");
+        java.util.stream.Stream.of(null, "", " ")
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri commented on cluster-operator/src/test/java/io/strimzi/operator/cluster/model/ModelUtilsTest.java:
```diff
@@ -181,4 +182,89 @@ public void testServiceDnsNames() {
+
+    @ParallelTest
+    public void testCreateOrAddConfigListWithNullValue() {
+        KafkaConfiguration config = new KafkaConfiguration(io.strimzi.operator.common.Reconciliation.DUMMY_RECONCILIATION, java.util.Set.of());
```
Why do you use the FQCN here?

OwenCorrigan76 replied:
Fixed

fvaleri reviewed (commented):
I left quite a few comments that will hopefully spare a lot more from the community. Also, you may have some IDE config that adds the FQCN to some of your code, I found a few but you should remove all of them and use imports instead.

---

We also have to enable MM2 connectors' metrics, as they are useful and used by the example Grafana dashboards. For this, you have to set the SMR configuration in `KafkaMirrorMaker2Connectors.prepareMirrorMaker2ConnectorConfig`, just before the return statement.

```java
// metrics are collected into a single registry and exposed using the Kafka Connect listener,
        // so we need to disable the listener here to avoid opening the default port (8080)
        if (strimziMetricsReporterEnabled) {
            config.put("metric.reporters", StrimziMetricsReporterConfig.KAFKA_CLASS);
            config.put(StrimziMetricsReporterConfig.LISTENER_ENABLE, "false");
        }
```

---

Once the Kafka PR is merged, you can also rebase this PR and introduce `StrimziMetricsReporterConfig` in `model/metrics`. There is no rush though, do it only when you feel the PR is ready for upstream, as it requires quite a few replacements.

```java
/**
 * This class holds Strimzi Metrics Reporter and related configuration keys.
 */
public class StrimziMetricsReporterConfig {
    /** KafkaMetricsReporter fully qualified class name. */
    public static final String KAFKA_CLASS = "io.strimzi.kafka.metrics.KafkaPrometheusMetricsReporter";

    /** YammerMetricsReporter fully qualified class name. */
    public static final String YAMMER_CLASS = "io.strimzi.kafka.metrics.YammerPrometheusMetricsReporter";
    
    /** Enable HTTP listener (true/false). */
    public static final String LISTENER_ENABLE = "prometheus.metrics.reporter.listener.enable";

    /** Set listener endpoint (e.g. http://:8080). */
    public static final String LISTENER = "prometheus.metrics.reporter.listener";

    /** Set a comma separated list of regexes. */
    public static final String ALLOW_LIST = "prometheus.metrics.reporter.allowlist";
}
```

StevenJDH committed:
Fix MirrorMaker 2 client rack init container override (#11429)

Signed-off-by: Steven Jenkins De Haro <20492442+StevenJDH@users.noreply.github.com>

OwenCorrigan76 committed:
Add support for the Strimzi Metrics Reporter to Kafka brokers/controllers components  (#11051)

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

see-quick committed:
chore: Minor formatting and style cleanups in ClusterOperator (#11436)

Signed-off-by: see-quick <maros.orsak159@gmail.com>

OwenCorrigan76 committed:
Add support for the Strimzi Metrics Reporter to brokers and controllers

This patch adds support for the Strimzi Metrics Reporter to brokers and controllers as described by the following proposal:

https://github.com/strimzi/proposals/blob/main/064-prometheus-metrics-reporter.md

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Squash commits

Address initial feedback

Signed-off-by: ocorriga <ocorriga@redhat.com>

Reverse dashboards and fix build issue

Signed-off-by: ocorriga <ocorriga@redhat.com>

Update derived resources

Signed-off-by: ocorriga <ocorriga@redhat.com>

Trigger build

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

Add MetricsModel interface and refactor affected classes

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Add CEL validation and address comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Addressing latest comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

add TODO in withStrimziMetricsReporter

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

address latest comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Add StrimziMetricsReporterConfig class

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Add defaults for allowlist

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Refactor withUserConfiguration and add tests

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address Comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Revert Connect script ENV VAR

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address Comments on common-configuration doc and fix typos

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Fix ConfigMap naming for tests

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address additional Comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address Jakub's and Tina's Comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Add comments to distinguish between kafka and yammer metrics

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address Paolo's comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Remove added comment writer.println

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Refactor behaviour of user config in KafkaBrokerConfigurationBuilder

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address Fede's comments

Signed-off-by: ocorriga <ocorriga@redhat.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address Mickael's comments

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Fix Kafka dashboard

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Update API doc

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Add system test

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Minor ST improvements

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>
Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address Kate's and Paolo's comments

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Remove jayway-json-path dependency from kafka-thirdparty-libs and add listener.enabed=true

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address nit and move prometheus-jmx-exporter examples into it's own directory

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address ST comments

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

Address Jakub's latest comments

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address comments on KafkaBrokerConfigurationBuilder

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Address comments from Jakub

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
Refactor to address Jmx Reporter change

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
Improve tests in KafkaBrokerConfigurationBuilderTest

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

OwenCorrigan76 committed:
More KafkaConnect refactor

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

PaulRMellor committed:
docs(tiered storage): updates storage considerations (#11444)

Signed-off-by: prmellor <pmellor@redhat.com>

yasiribmcon committed:
Provide full qualified name for KubeRegistryProxy for s390x (#11445)

Signed-off-by: Yasir Ashfaq <Yasir.Ashfaq1@ibm.com>

OwenCorrigan76 committed:
Update assemblies and modules Docs for Metrics Reporter integration (#11329)

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

Frawless committed:
Enhance CI scripts to better work with arm runners and bump Minikube version (#11443)

Signed-off-by: Jakub Stejskal <xstejs24@gmail.com>

Frawless committed:
[ST] Bump Jaeger version to 1.65.0 to make system tests working on Kube 1.33 (#11460)

Signed-off-by: Jakub Stejskal <xstejs24@gmail.com>

PaulRMellor committed:
docs(fix): fixes some broken links in the doc (#11454)

Signed-off-by: prmellor <pmellor@redhat.com>

im-konge committed:
Update Minikube version to 1.36.0 (#11474)

see-quick committed:
[system test] Update TESTING.md doc (#11471)

Signed-off-by: see-quick <maros.orsak159@gmail.com>

ppatierno committed:
Added support for Apache Kafka 3.9.1 (#11469)

Signed-off-by: Paolo Patierno <ppatierno@live.com>

Frawless committed:
[ST] Update Cert Manager to 1.17 and Jaeger to 2.6 in system tests (#11479)

Signed-off-by: Jakub Stejskal <xstejs24@gmail.com>

ppatierno committed:
Update to Vert.x 5 (#11374)

Signed-off-by: Paolo Patierno <ppatierno@live.com>

ppatierno committed:
Fixed adding new Kafka versions documentation (#11482)

Signed-off-by: Paolo Patierno <ppatierno@live.com>

see-quick committed:
[system test] [doc] MirrorMaker2ST (#11473)

Signed-off-by: see-quick <maros.orsak159@gmail.com>
Signed-off-by: Maros Orsak <maros.orsak159@gmail.com>
Co-authored-by: Jakub Stejskal <xstejs24@gmail.com>

Frawless committed:
Allow us to change control and worker nodes count via env vars, bump Kind version (#11458)

Signed-off-by: Jakub Stejskal <xstejs24@gmail.com>

scholzj committed:
Update different dependencies - Kaniko, Fabric8, JMX Exporter, Maven builder (#11492)

Signed-off-by: Jakub Scholz <www@scholzj.com>

ppatierno committed:
Moved to use the Sonatype Central Portal (#11493)

Signed-off-by: Paolo Patierno <ppatierno@live.com>

Frawless committed:
Use sudo for docker manipulation in Kind script (#11497)

Signed-off-by: Jakub Stejskal <xstejs24@gmail.com>

im-konge committed:
Add note/warning about Kubernetes 1.25 and 1.26 support removal (#11498)

Signed-off-by: Lukas Kral <lukywill16@gmail.com>

ppatierno committed:
Added project names for the ones to be published to Maven Central (#11501)

Signed-off-by: Paolo Patierno <ppatierno@live.com>

scholzj committed:
Add support for Image Volumes (#11467)

Signed-off-by: Jakub Scholz <www@scholzj.com>

PaulRMellor committed:
docs(kraft): clarifies static quorum limit applies to new deployments (#11505)

Signed-off-by: prmellor <pmellor@redhat.com>

Frawless committed:
[ST] Add systemtest with user specific configuraton for ssl.* fields for custom type of listener (#11500)

Signed-off-by: Jakub Stejskal <xstejs24@gmail.com>

OwenCorrigan76 committed:
Address initial comments

Signed-off-by: OwenCorrigan76 <owencorrigan76@gmail.com>

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaMirrorMaker2Connectors.java:
```diff
@@ -83,6 +85,7 @@ private KafkaMirrorMaker2Connectors(Reconciliation reconciliation, KafkaMirrorMa
         this.mirrors = kafkaMirrorMaker2.getSpec().getMirrors();
         this.tracing = kafkaMirrorMaker2.getSpec().getTracing();
         this.rackAwarenessEnabled = kafkaMirrorMaker2.getSpec().getRack() != null;
+        //this.strimziMetricsReporterEnabled = kafkaMirrorMaker2.getSpec().getMetricsConfig() != null;
```
You have to use `instanceof`, because it could be the other metrics type. Don't forget to enable all related code.

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -269,20 +279,51 @@ private void printConfigProviders(AbstractConfiguration userConfig) {
 
+    private void maybeAddMetricReporters(AbstractConfiguration userConfig, boolean injectKafkaJmxReporter, boolean injectStrimziMetricsReporter) {
+        if (injectKafkaJmxReporter) {
+            String values = "org.apache.kafka.common.metrics.JmxReporter";
```
Add the following comment before this line:

```
// the JMX reporter needs to be explicitly added when metric.reporters is not empty
```

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfigurationBuilder.java:
```diff
@@ -269,20 +279,51 @@ private void printConfigProviders(AbstractConfiguration userConfig) {
+            ModelUtils.createOrAddListConfig(userConfig, "consumer.metric.reporters", values);
+        }
+        if (injectStrimziMetricsReporter) {
+            String values = "io.strimzi.kafka.metrics.KafkaPrometheusMetricsReporter";
```
Add the following comment before this line:

```
// we also need to explicitly configure Kafka clients in this case
```

fvaleri commented on cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaConnectConfiguration.java:
```diff
@@ -34,13 +34,25 @@ public class KafkaConnectConfiguration extends AbstractConfiguration {
      * @param jsonOptions     Json object with configuration options as key ad value pairs.
      */
-    public KafkaConnectConfiguration(Reconciliation reconciliation, Iterable<Map.Entry<String, Object>> jsonOptions) {
+    public KafkaConnectConfiguration(Iterable<Map.Entry<String, Object>> jsonOptions, Reconciliation reconciliation) {
```
Can we have `reconciliation` as first parameter like in `KafkaMirrorMaker2Configuration` and `KafkaConfiguration`?

fvaleri reviewed (commented):
I left some comments, but there are a lot of unrelated changes and it's kind hard to do a proper review. Did you rebased against main branch?
